// @microsoft/bot-components-teams/lib/actions/getTeamMember.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetTeamMember = void 0;
const adaptive_expressions_1 = require("adaptive-expressions");
const botbuilder_1 = require("botbuilder");
const botbuilder_dialogs_1 = require("botbuilder-dialogs");
const actionHelpers_1 = require("./actionHelpers");
/**
 * Calls TeamsInfo.GetTeamMember to retrieve the list of members in a teams
 * scoped conversation and sets the result to a memory property.
 */
class GetTeamMember extends botbuilder_dialogs_1.Dialog {
    constructor() {
        super(...arguments);
        /**
         * Gets or sets the expression to get the value to use for member id.
         *
         * @default
         * =turn.activity.from.id
         */
        this.memberId = new adaptive_expressions_1.StringExpression('=turn.activity.from.id');
        /**
         * Gets or sets the expression to get the value to use for team id.
         *
         * @default
         * =turn.activity.channelData.team.id
         */
        this.teamId = new adaptive_expressions_1.StringExpression('=turn.activity.channelData.team.id');
    }
    getConverter(property) {
        switch (property) {
            case 'disabled':
                return new adaptive_expressions_1.BoolExpressionConverter();
            case 'property':
            case 'memberId':
            case 'teamId':
                return new adaptive_expressions_1.StringExpressionConverter();
            default:
                return super.getConverter(property);
        }
    }
    /**
     * Called when the dialog is started and pushed onto the dialog stack.
     *
     * @param {DialogContext} dc The [DialogContext](xref:botbuilder-dialogs.DialogContext) for the current turn of conversation.
     * @param {object} _options Optional, initial information to pass to the dialog.
     * @returns {Promise<DialogTurnResult>} A promise representing the asynchronous operation.
     */
    beginDialog(dc, _options) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function* () {
            if ((_a = this.disabled) === null || _a === void 0 ? void 0 : _a.getValue(dc.state)) {
                return dc.endDialog();
            }
            if (dc.context.activity.channelId !== botbuilder_1.Channels.Msteams) {
                throw new Error(`${GetTeamMember.$kind} works only on the Teams channel.`);
            }
            const memberId = (0, actionHelpers_1.getValue)(dc, this.memberId);
            if (!memberId) {
                throw new Error(`Missing MemberId in ${GetTeamMember.$kind}.`);
            }
            const teamId = (0, actionHelpers_1.getValue)(dc, this.teamId);
            const result = yield botbuilder_1.TeamsInfo.getTeamMember(dc.context, teamId, memberId);
            if (this.property != null) {
                dc.state.setValue((_b = this.property) === null || _b === void 0 ? void 0 : _b.getValue(dc.state), result);
            }
            return dc.endDialog(result);
        });
    }
    /**
     * Builds the compute Id for the dialog.
     *
     * @returns {string} A string representing the compute Id.
     */
    onComputeId() {
        var _a, _b, _c, _d, _e, _f;
        return `GetTeamMember[\
            ${(_b = (_a = this.memberId) === null || _a === void 0 ? void 0 : _a.toString()) !== null && _b !== void 0 ? _b : ''},\
            ${(_d = (_c = this.teamId) === null || _c === void 0 ? void 0 : _c.toString()) !== null && _d !== void 0 ? _d : ''},\
            ${(_f = (_e = this.property) === null || _e === void 0 ? void 0 : _e.toString()) !== null && _f !== void 0 ? _f : ''}\
        ]`;
    }
}
exports.GetTeamMember = GetTeamMember;
/**
 * Class identifier.
 */
GetTeamMember.$kind = 'Teams.GetTeamMember';
