// @microsoft/bot-components-teams/lib/actions/sendMEConfigQuerySettingUrlResponse.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SendMEConfigQuerySettingUrlResponse = void 0;
const adaptive_expressions_1 = require("adaptive-expressions");
const botbuilder_1 = require("botbuilder");
const baseTeamsCacheInfoResponseDialog_1 = require("./baseTeamsCacheInfoResponseDialog");
/**
 * Send a messaging extension 'config' response. This is the type of response used for a 'composeExtension/querySettingUrl' request.
 */
class SendMEConfigQuerySettingUrlResponse extends baseTeamsCacheInfoResponseDialog_1.BaseTeamsCacheInfoResponseDialog {
    getConverter(property) {
        switch (property) {
            case 'disabled':
                return new adaptive_expressions_1.BoolExpressionConverter();
            case 'configUrl':
                return new adaptive_expressions_1.StringExpressionConverter();
            default:
                return super.getConverter(property);
        }
    }
    /**
     * Called when the dialog is started and pushed onto the dialog stack.
     *
     * @param {DialogContext} dc The [DialogContext](xref:botbuilder-dialogs.DialogContext) for the current turn of conversation.
     * @param {object} _options Optional, initial information to pass to the dialog.
     * @returns {Promise<DialogTurnResult>} A promise representing the asynchronous operation.
     */
    beginDialog(dc, _options) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function* () {
            if ((_a = this.disabled) === null || _a === void 0 ? void 0 : _a.getValue(dc.state)) {
                return dc.endDialog();
            }
            const configUrl = (_b = this.configUrl) === null || _b === void 0 ? void 0 : _b.getValue(dc.state);
            if (!configUrl) {
                throw new Error(`ConfigUrl is required for ${SendMEConfigQuerySettingUrlResponse.$kind}.`);
            }
            const response = {
                composeExtension: {
                    type: 'config',
                    suggestedActions: {
                        actions: [
                            {
                                type: botbuilder_1.ActionTypes.OpenUrl,
                                value: encodeURI(configUrl),
                                title: '', // title is never seen by the user but is a required property in CardAction.
                            },
                        ],
                    },
                },
                cacheInfo: this.getCacheInfo(dc),
            };
            const invokeResponse = baseTeamsCacheInfoResponseDialog_1.BaseTeamsCacheInfoResponseDialog.createInvokeResponseActivity(response);
            const resourceResponse = yield dc.context.sendActivity(invokeResponse);
            return dc.endDialog(resourceResponse);
        });
    }
    /**
     * Builds the compute Id for the dialog.
     *
     * @returns {string} A string representing the compute Id.
     */
    onComputeId() {
        var _a, _b;
        return `SendMEConfigQuerySettingsUrlResponse[\
            ${botbuilder_1.StringUtils.ellipsis((_b = (_a = this.configUrl) === null || _a === void 0 ? void 0 : _a.toString()) !== null && _b !== void 0 ? _b : '', 20)}\
        ]`;
    }
}
exports.SendMEConfigQuerySettingUrlResponse = SendMEConfigQuerySettingUrlResponse;
/**
 * Class identifier.
 */
SendMEConfigQuerySettingUrlResponse.$kind = 'Teams.SendMEConfigQuerySettingUrlResponse';
