// @microsoft/bot-components-teams/lib/actions/sendMEAttachmentsResponse.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SendMEAttachmentsResponse = void 0;
const adaptive_expressions_1 = require("adaptive-expressions");
const botbuilder_1 = require("botbuilder");
const converters_1 = require("botbuilder-dialogs-adaptive/lib/converters");
const baseTeamsCacheInfoResponseDialog_1 = require("./baseTeamsCacheInfoResponseDialog");
/**
 * Send a messaging extension 'result' response when a Teams Invoke Activity is received with activity.name='composeExtension/queryLink'.
 */
class SendMEAttachmentsResponse extends baseTeamsCacheInfoResponseDialog_1.BaseTeamsCacheInfoResponseDialog {
    constructor() {
        super(...arguments);
        /**
         * Gets or sets the Attachment Layout type for the response ('grid' or 'list').
         *
         * @default
         * list
         */
        this.attachmentLayout = new adaptive_expressions_1.EnumExpression('list');
    }
    getConverter(property) {
        switch (property) {
            case 'property':
                return new adaptive_expressions_1.StringExpressionConverter();
            case 'attachments':
                return new converters_1.ActivityTemplateConverter();
            case 'attachmentLayout':
                return new adaptive_expressions_1.EnumExpressionConverter(this.attachmentLayout);
            default:
                return super.getConverter(property);
        }
    }
    /**
     * Called when the dialog is started and pushed onto the dialog stack.
     *
     * @param {DialogContext} dc The [DialogContext](xref:botbuilder-dialogs.DialogContext) for the current turn of conversation.
     * @param {object} _options Optional, initial information to pass to the dialog.
     * @returns {Promise<DialogTurnResult>} A promise representing the asynchronous operation.
     */
    beginDialog(dc, _options) {
        var _a, _b, _c, _d;
        return __awaiter(this, void 0, void 0, function* () {
            if ((_a = this.disabled) === null || _a === void 0 ? void 0 : _a.getValue(dc.state)) {
                return dc.endDialog();
            }
            const activity = yield ((_b = this.attachments) === null || _b === void 0 ? void 0 : _b.bind(dc, dc.state));
            if (!((_c = activity === null || activity === void 0 ? void 0 : activity.attachments) === null || _c === void 0 ? void 0 : _c.length)) {
                throw new Error(`Missing attachments in ${SendMEAttachmentsResponse.$kind}.`);
            }
            const layout = (_d = this.attachmentLayout) === null || _d === void 0 ? void 0 : _d.getValue(dc.state);
            const result = {
                type: 'result',
                attachmentLayout: layout,
                attachments: activity.attachments,
            };
            const invokeResponse = this.createMessagingExtensionInvokeResponseActivity(dc, result);
            const resourceResponse = yield dc.context.sendActivity(invokeResponse);
            return dc.endDialog(resourceResponse);
        });
    }
    /**
     * Builds the compute Id for the dialog.
     *
     * @returns {string} A string representing the compute Id.
     */
    onComputeId() {
        var _a, _b;
        return `SendMEAttachmentsResponse[\
            ${botbuilder_1.StringUtils.ellipsisHash((_b = (_a = this.attachments) === null || _a === void 0 ? void 0 : _a.toString()) !== null && _b !== void 0 ? _b : '', 20)}\
        ]`;
    }
}
exports.SendMEAttachmentsResponse = SendMEAttachmentsResponse;
/**
 * Class identifier.
 */
SendMEAttachmentsResponse.$kind = 'Teams.SendMEAttachmentsResponse';
