// @microsoft/bot-components-teams/lib/actions/sendTabCardResponse.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SendTabCardResponse = void 0;
const adaptive_expressions_1 = require("adaptive-expressions");
const botbuilder_1 = require("botbuilder");
const botbuilder_dialogs_1 = require("botbuilder-dialogs");
const converters_1 = require("botbuilder-dialogs-adaptive/lib/converters");
/**
 * Send a Card Tab response to the user.
 */
class SendTabCardResponse extends botbuilder_dialogs_1.Dialog {
    getConverter(property) {
        switch (property) {
            case 'disabled':
                return new adaptive_expressions_1.BoolExpressionConverter();
            case 'property':
                return new adaptive_expressions_1.StringExpressionConverter();
            case 'cards':
                return new converters_1.ActivityTemplateConverter();
            default:
                return super.getConverter(property);
        }
    }
    /**
     * Called when the dialog is started and pushed onto the dialog stack.
     *
     * @param {DialogContext} dc The [DialogContext](xref:botbuilder-dialogs.DialogContext) for the current turn of conversation.
     * @param {object} _options Optional, initial information to pass to the dialog.
     * @returns {Promise<DialogTurnResult>} A promise representing the asynchronous operation.
     */
    beginDialog(dc, _options) {
        var _a, _b, _c;
        return __awaiter(this, void 0, void 0, function* () {
            if ((_a = this.disabled) === null || _a === void 0 ? void 0 : _a.getValue(dc.state)) {
                return dc.endDialog();
            }
            if (!this.cards) {
                throw new Error(`Valid Cards are required for ${SendTabCardResponse.$kind}.`);
            }
            const activity = yield ((_b = this.cards) === null || _b === void 0 ? void 0 : _b.bind(dc, dc.state));
            if (!((_c = activity === null || activity === void 0 ? void 0 : activity.attachments) === null || _c === void 0 ? void 0 : _c.length)) {
                throw new Error(`Invalid activity. Attachment(s) are required for ${SendTabCardResponse.$kind}.`);
            }
            const cards = activity.attachments.map((attachment) => {
                return {
                    card: attachment.content,
                };
            });
            const responseActivity = this.getTabInvokeResponse(cards);
            const sendResponse = yield dc.context.sendActivity(responseActivity);
            return dc.endDialog(sendResponse);
        });
    }
    /**
     * Builds the compute Id for the dialog.
     *
     * @returns {string} A string representing the compute Id.
     */
    onComputeId() {
        var _a, _b;
        return `SendTabCardResponse[\
            ${(_b = (_a = this.cards) === null || _a === void 0 ? void 0 : _a.toString()) !== null && _b !== void 0 ? _b : ''}\
        ]`;
    }
    getTabInvokeResponse(cards) {
        return {
            value: {
                status: botbuilder_1.StatusCodes.OK,
                body: {
                    tab: {
                        type: 'continue',
                        value: {
                            cards,
                        },
                    },
                },
            },
            type: botbuilder_1.ActivityTypes.InvokeResponse,
        };
    }
}
exports.SendTabCardResponse = SendTabCardResponse;
/**
 * Class identifier.
 */
SendTabCardResponse.$kind = 'Teams.SendTabCardResponse';
