// @microsoft/bot-components-teams/lib/actions/sendTabAuthResponse.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.SendTabAuthResponse = void 0;
const baseAuthResponseDialog_1 = require("./baseAuthResponseDialog");
const baseTeamsCacheInfoResponseDialog_1 = require("./baseTeamsCacheInfoResponseDialog");
/**
 * Send a tab 'auth' response.
 */
class SendTabAuthResponse extends baseAuthResponseDialog_1.BaseAuthResponseDialog {
    /**
     * Builds the compute Id for the dialog.
     *
     * @returns {string} A string representing the compute Id.
     */
    onComputeId() {
        var _a, _b;
        return `SendTabAuthResponse[\
            ${(_b = (_a = this.title) === null || _a === void 0 ? void 0 : _a.toString()) !== null && _b !== void 0 ? _b : ''}\
        ]`;
    }
    createOAuthInvokeResponseActivityFromCardAction(dc, cardAction) {
        const responsePayload = {
            type: 'auth',
            suggestedActions: {
                actions: [cardAction],
            },
        };
        const tabResponse = { tab: responsePayload };
        return baseTeamsCacheInfoResponseDialog_1.BaseTeamsCacheInfoResponseDialog.createInvokeResponseActivity(tabResponse);
    }
}
exports.SendTabAuthResponse = SendTabAuthResponse;
/**
 * Class identifier.
 */
SendTabAuthResponse.$kind = 'Teams.SendTabAuthResponse';
