// @microsoft/bot-components-teams/lib/actions/sendTaskModuleUrlResponse.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SendTaskModuleUrlResponse = void 0;
const adaptive_expressions_1 = require("adaptive-expressions");
const botbuilder_1 = require("botbuilder");
const baseSendTaskModuleContinueResponse_1 = require("./baseSendTaskModuleContinueResponse");
const baseTeamsCacheInfoResponseDialog_1 = require("./baseTeamsCacheInfoResponseDialog");
/**
 * Send a simple message task module response.
 */
class SendTaskModuleUrlResponse extends baseSendTaskModuleContinueResponse_1.BaseSendTaskModuleContinueResponse {
    getConverter(property) {
        switch (property) {
            case 'disabled':
                return new adaptive_expressions_1.BoolExpressionConverter();
            case 'url':
            case 'fallbackUrl':
                return new adaptive_expressions_1.StringExpressionConverter();
            default:
                return super.getConverter(property);
        }
    }
    /**
     * Called when the dialog is started and pushed onto the dialog stack.
     *
     * @param {DialogContext} dc The [DialogContext](xref:botbuilder-dialogs.DialogContext) for the current turn of conversation.
     * @param {object} _options Optional, initial information to pass to the dialog.
     * @returns {Promise<DialogTurnResult>} A promise representing the asynchronous operation.
     */
    beginDialog(dc, _options) {
        var _a, _b, _c, _d, _e, _f, _g;
        return __awaiter(this, void 0, void 0, function* () {
            if ((_a = this.disabled) === null || _a === void 0 ? void 0 : _a.getValue(dc.state)) {
                return dc.endDialog();
            }
            const url = (_b = this.url) === null || _b === void 0 ? void 0 : _b.getValue(dc.state);
            if (!url) {
                throw new Error(`Missing Url for ${SendTaskModuleUrlResponse.$kind}.`);
            }
            const title = (_c = this.title) === null || _c === void 0 ? void 0 : _c.getValue(dc.state);
            const height = (_d = this.height) === null || _d === void 0 ? void 0 : _d.getValue(dc.state);
            const width = (_e = this.width) === null || _e === void 0 ? void 0 : _e.getValue(dc.state);
            const completionBotId = (_f = this.completionBotId) === null || _f === void 0 ? void 0 : _f.getValue(dc.state);
            const fallbackUrl = (_g = this.fallbackUrl) === null || _g === void 0 ? void 0 : _g.getValue(dc.state);
            const response = {
                task: {
                    type: 'continue',
                    value: {
                        height,
                        width,
                        title,
                        completionBotId,
                        url: encodeURI(url),
                        fallbackUrl: encodeURI(fallbackUrl || ''),
                    },
                },
                cacheInfo: this.getCacheInfo(dc),
            };
            const responseActivity = baseTeamsCacheInfoResponseDialog_1.BaseTeamsCacheInfoResponseDialog.createInvokeResponseActivity(response);
            const sendResponse = yield dc.context.sendActivity(responseActivity);
            return dc.endDialog(sendResponse);
        });
    }
    /**
     * Builds the compute Id for the dialog.
     *
     * @returns {string} A string representing the compute Id.
     */
    onComputeId() {
        var _a, _b, _c, _d;
        return `SendTaskModuleUrlResponse[\
            ${botbuilder_1.StringUtils.ellipsis((_b = (_a = this.url) === null || _a === void 0 ? void 0 : _a.toString()) !== null && _b !== void 0 ? _b : '', 20)},\
            ${(_d = (_c = this.title) === null || _c === void 0 ? void 0 : _c.toString()) !== null && _d !== void 0 ? _d : ''}\
        ]`;
    }
}
exports.SendTaskModuleUrlResponse = SendTaskModuleUrlResponse;
/**
 * Class identifier.
 */
SendTaskModuleUrlResponse.$kind = 'Teams.SendTaskModuleUrlResponse';
