// @microsoft/bot-components-teams/lib/actions/sendMEAuthResponse.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.SendMEAuthResponse = void 0;
const baseAuthResponseDialog_1 = require("./baseAuthResponseDialog");
/**
 * Send a messaging extension 'result' response when a Teams Invoke Activity is received with activity.name='composeExtension/queryLink'.
 */
class SendMEAuthResponse extends baseAuthResponseDialog_1.BaseAuthResponseDialog {
    /**
     * Builds the compute Id for the dialog.
     *
     * @returns {string} A string representing the compute Id.
     */
    onComputeId() {
        var _a, _b;
        return `SendMessagingExtensionAuthResponse[\
            ${(_b = (_a = this.title) === null || _a === void 0 ? void 0 : _a.toString()) !== null && _b !== void 0 ? _b : ''}\
        ]`;
    }
    createOAuthInvokeResponseActivityFromCardAction(dc, cardAction) {
        const result = {
            type: 'auth',
            suggestedActions: {
                actions: [cardAction],
            },
        };
        return this.createMessagingExtensionInvokeResponseActivity(dc, result);
    }
}
exports.SendMEAuthResponse = SendMEAuthResponse;
/**
 * Class identifier.
 */
SendMEAuthResponse.$kind = 'Teams.SendMEAuthResponse';
